### DISCUSSION TITLE
Add "call-by-need" parameters evaluation

### DISCUSSION THREAD

**Discussion by riiswa** [scala/scala3#17881] [Feature Requests]
Consider the following function: 

```scala
def f(x: Int, lazy y: Int): Int = 
  if (x%2==0) x else y+y
```
With the lazy parameters, if we pass in parameter y the following block: `{println("Hello World"); 1}`, it will be evaluated 0 or 1 time, unlike the call-by-value which evaluates 1 time and the call-by-name which evaluates 0 or 2 times.

This behavior is possible in Scala in the following way: 
```scala
def f(x: Int, lazy y: => Int): Int ={ 
  lazy val y_ = y
  if (x%2==0) x else y_+y_
}
```

I think it would be useful to have in Scala 3 the keyword lazy for function parameters.

**Comment by som-snytt** [comment]
https://github.com/scala/bug/issues/240

**Comment by som-snytt** [comment]
or https://github.com/lampepfl/dotty-feature-requests/issues/11
